subprojects/publish/src/main/java/org/gradle/api/publish/internal/mapping/VariantDependencyResolver.java

```diff
@@ -0,0 +1,115 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.publish.internal.mapping;
+
+import org.gradle.api.artifacts.DependencyConstraint;
+import org.gradle.api.artifacts.ModuleDependency;
+import org.gradle.api.artifacts.ModuleVersionIdentifier;
+import org.gradle.api.publish.internal.validation.VariantWarningCollector;
+
+import javax.annotation.Nullable;
+
+/**
+ * Given declared dependencies and dependency constraints, determines the coordinates
+ * that should be used to reference the dependency in published metadata. Instances
+ * of this interface are scoped to a given published variant, as each variant may
+ * have different rules for resolving published dependencies.
+ *
+ * <p>This resolver can resolve dependencies to either component or variant precision.
+ * When resolving to component precision, the returned coordinates will reference the
+ * root coordinates of a multi-coordinate publication. When resolving to variant precision,
+ * the returned coordinates will reference the child coordinates of the variant selected by
+ * the resolved dependency.</p>
+ */
+public interface VariantDependencyResolver {
+
+    /**
+     * Determines the published coordinates for a dependency to variant-level precision.
+     *
+     * @throws RuntimeException If {@code dependency} is a project dependency and the project cannot be resolved.
+     */
+    Coordinates resolveVariantCoordinates(ModuleDependency dependency, VariantWarningCollector warnings);
+
+    /**
+     * Determines the published coordinates for a dependency constraint to variant-level precision.
+     *
+     * @throws RuntimeException If {@code dependency} is a project dependency constraint and the project cannot be resolved.
+     */
+    Coordinates resolveVariantCoordinates(DependencyConstraint dependency, VariantWarningCollector warnings);
+
+    /**
+     * Determines the published coordinates for a dependency to component-level precision.
+     *
+     * @throws RuntimeException If {@code dependency} is a project dependency and the project cannot be resolved.
+     */
+    Coordinates resolveComponentCoordinates(ModuleDependency dependency);
+
+    /**
+     * Determines the published coordinates for a dependency constraint to component-level precision.
+     *
+     * @throws RuntimeException If {@code dependency} is a project dependency constraint and the project cannot be resolved.
+     */
+    Coordinates resolveComponentCoordinates(DependencyConstraint dependency);
+
+    /**
+     * Similar to {@link ModuleVersionIdentifier}, but allows a null version.
+     */
+    interface Coordinates {
```

Comment:
💭 I think this should just be a simple value class with 2 constructors.  I'd name it something more specific, too.  `PublishedCoordinates` perhaps?

Reply:
Renamed to `ResolvedCoordinates`

The intention of having two implementations was to optimize the `ModuleVersionIdentifier` factory method to avoid copying all 3 GAV fields into a new `ResolvedCoordinates` instance. During publishing a lot of these instances are created and having 1 field > having 3 fields

Perhaps this is an unnecessary optimization, but the `ResolvedCoordinates` class is relatively self-contained. IMO its fine the way it is

Reply:
That makes sense.  Can you add it as Javadoc on the type?

Reply:
It will be there on the follow up PR which is ready to push up when this goes though